### PR TITLE
Task: Spring Boot Portfolio Project — Part 6. Executor: Oleksii Bashkirov.

### PR DESCRIPTION
**Pull request by OleksiiBashkirov**
feat: Incorporate Spring Security with Role-based Access Control

- Added `spring-boot-starter-security` to the project.
- Created `SecurityConfig` class with `PasswordEncoder` and `SecurityFilterChain` beans.
- Implemented `UserDetails` in `User` class and `UserDetailsService`.
- Made `/auth/registration` publicly available.
- Added `Role` entity and `RoleRepository`.
- Added `Set<Role> roles` field to `User` entity.
- Updated Liquibase changesets for roles and user roles.
- Secured endpoints with `@PreAuthorize` based on roles.

### CODE REVIEW

File: src/main/java/mate/academy/repository/RoleRepository.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package mate.academy.repository;
+
+import mate.academy.model.Role;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/repository/RoleRepository.java]:
```suggestion
```

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 2]
```diff
@@ -101,6 +101,12 @@
             <artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
             <version>2.1.0</version>
         </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-security</artifactId>
+            <version>3.3.0</version>
```
Comment (Elena-Bruyako) [inline pom.xml]:
```suggestion
```

File: src/main/java/mate/academy/config/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.http.HttpMethod;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@Configuration
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/config/SecurityConfig.java:17]:
```suggestion
@Configuration
@EnableMethodSecurity
```

File: src/main/java/mate/academy/config/SecurityConfig.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            .cors().disable()
+            .csrf().disable()
+            .authorizeHttpRequests(auth -> auth
+                .requestMatchers("/auth/**", "/error", "/swagger-ui/**").permitAll()
+                .requestMatchers(HttpMethod.GET, "/api/books/**").hasRole("USER")
+                .requestMatchers(HttpMethod.POST, "/api/books/**").hasRole("ADMIN")
+                .requestMatchers(HttpMethod.PUT, "/api/books/**").hasRole("ADMIN")
+                .requestMatchers(HttpMethod.DELETE, "/api/books/**").hasRole("ADMIN")
+                .anyRequest().authenticated()
+            )
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/config/SecurityConfig.java]: add userDetailsService

File: src/main/java/mate/academy/model/User.java [hunk 5]
```diff
@@ -31,4 +40,47 @@ public class User {
     private String lastName;
 
     private String shippingAddress;
+
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/model/User.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/security/CustomUserDetailsService.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+@RequiredArgsConstructor
+@Service
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String username)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(username)
+                .orElseThrow(() -> new UsernameNotFoundException("User not found."));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/security/CustomUserDetailsService.java]:
```suggestion
{
        return userRepository.findByEmail(username)
                .orElseThrow(() -> new EntityNotFoundException("User not found."));
```

File: src/main/java/mate/academy/service/impl/UserServiceImpl.java [hunk 7]
```diff
@@ -24,6 +26,7 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         }
 
         User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/UserServiceImpl.java:33]: find role by name add it to roles

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/security/CustomUserDetailsService.java [hunk 8]
```diff
@@ -14,8 +19,22 @@ public class CustomUserDetailsService implements UserDetailsService {
...
-            throws UsernameNotFoundException {
-        return userRepository.findByEmail(username)
-                .orElseThrow(() -> new UsernameNotFoundException("User not found."));
+            throws EntityNotFoundException {
+
+        User user = userRepository.findByEmail(username)
+                .orElseThrow(() -> new UsernameNotFoundException("User not found: " + username));
+
+        System.out.println("User found: " + user.getEmail());
+        System.out.println("User roles: " + user.getRoles());
```
Comment (sarakhmen) [inline src/main/java/mate/academy/security/CustomUserDetailsService.java]: remove these debug messages

**Review by sarakhmen — CHANGES_REQUESTED** [review]
let`s improve your sokution

File: src/main/java/mate/academy/security/CustomUserDetailsService.java (hunk 8 above)
Comment (OleksiiBashkirov) [reply src/main/java/mate/academy/security/CustomUserDetailsService.java]: Thank you for pointing that out. 
I had accidentally left those debug messages in, but I have now removed them.
Thank you for your guidance!

File: src/main/java/mate/academy/config/SecurityConfig.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@Configuration
+@EnableWebSecurity
+@EnableMethodSecurity
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/config/SecurityConfig.java:18]: Do you need both of these annotations `@EnableWebSecurity and @EnableMethodSecurity`?

File: src/main/java/mate/academy/mapper/BookMapper.java [hunk 10]
```diff
@@ -6,7 +6,9 @@
 import mate.academy.model.Book;
 import org.mapstruct.Mapper;
 import org.mapstruct.MappingTarget;
+import org.springframework.stereotype.Component;
 
+@Component
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/mapper/BookMapper.java]:
```suggestion
```

File: src/main/java/mate/academy/model/User.java [hunk 11]
```diff
@@ -31,4 +39,47 @@ public class User {
     private String lastName;
 
     private String shippingAddress;
+
+    @ManyToMany()
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/model/User.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/security/CustomUserDetailsService.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+        List<SimpleGrantedAuthority> authorities = user.getRoles().stream()
+                .map(role -> new SimpleGrantedAuthority(role.getName().name()))
+                .collect(Collectors.toList());
+
+        return new org.springframework.security.core.userdetails.User(
+                user.getEmail(),
+                user.getPassword(),
+                authorities
+        );
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/security/CustomUserDetailsService.java]:
```suggestion
           {

        return userRepository.findByEmail(username)
                .orElseThrow(() -> new EntityNotFoundException("User not found: " + username));
```

File: src/main/java/mate/academy/service/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -24,6 +30,10 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
...
+        Role userRole = roleRepository.findByName(Role.RoleName.ROLE_USER)
+                        .orElseThrow(() -> new RegistrationException("Role not found."));
+        user.setRoles(Collections.singleton(userRole));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/UserServiceImpl.java]: or you can use Set.of()

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]
well done

**Comment by OleksiiBashkirov** [comment]
Please disregard the last pull request in this theme made through the mate.academy platform. 
Due to an accidental click, I submitted changes meant for the "JWT" theme under the "Spring Boot Security" theme. 
Sorry about that!

**Pull request merged**